pulser-core/pulser/pulse.py

```diff
@@ -182,6 +189,45 @@ def ConstantPulse(
         detuning_wf = ConstantWaveform(duration, detuning)
         return cls(amplitude_wf, detuning_wf, phase, post_phase_shift)
 
+    @classmethod
+    @parametrize
+    def ArbitraryPhase(
+        cls,
+        amplitude: Waveform | Parametrized,
+        phase: Waveform | Parametrized,
+        post_phase_shift: float | Parametrized = 0.0,
+    ) -> Pulse:
+        """Pulse with an arbitrary phase waveform.
+
+        The detuning waveform is the derivative of the phase waveform while the
+        phase offset is simply the phase at the beginning of the waveform. In
+        the detuning waveform calculation, it is always assumed that the
+        detuning is the same in the first two samples.
+
+        Args:
+            amplitude: The amplitude waveform (in rad/µs).
+            phase: The phase waveform (in rads).
+            post_phase_shift: Optionally lets you add a
+                phase shift (in rads) immediately after the end of the pulse.
+        """
+        if not isinstance(phase, Waveform):
+            raise TypeError(
+                f"'phase' must be a waveform, not of type {type(phase)}."
+            )
+        phase_0 = phase.first_value
+        detuning: Waveform
+        if isinstance(phase, ConstantWaveform):
+            detuning = ConstantWaveform(phase.duration, 0.0)
+        elif isinstance(phase, RampWaveform):
+            detuning = ConstantWaveform(phase.duration, phase.slope * 1e3)
+        else:
+            detuning_samples = np.diff(phase.samples) * 1e3  # rad/ns -> rad/µs
+            # Use the same value in the first two detuning samples
+            detuning = CustomWaveform(
+                np.pad(detuning_samples, (1, 0), mode="edge")
+            )
```

Comment:
I thought about that, but then it was not consistent with the special case of a RampWaveform. Plus, it will introduce a 1ns jump down to zero in the detuning

Reply:
Yeah I figured you had thought about that, I stand behind your conclusions. I have the feeling that this $\delta[0]$ is just adding a constant phase to the modulated phase, and so it should have a huge impact (I hope)

Reply:
> I don't think that it matters much, but from a mathematical point of view I think it would be better if the first detuning samples was 0: Σk=0N−1δ[k]=Σk=1N−1(ϕ[k]−ϕ[k−1])+δ[0] =Σk=1N−1ϕ[k]−Σk=1N−1ϕ[k−1]+δ[0] =Σk=1N−1ϕ[k]−Σk=0N−2ϕ[k]+δ[0] =ϕ[N−1]+δ[0] But that doesn't take into account modulation

Btw I found a mistake in this development:
Let's take $M\ge 1$:
$$\Sigma_{k=0}^{M} \delta\[k\] = \Sigma_{k=1}^{M} (\phi[k] - \phi[k-1]) + \delta[0] $$
$$= \Sigma_{k=1}^{M} \phi[k] - \Sigma_{k=1}^{M} \phi[k-1] + \delta[0]$$
$$= \Sigma_{k=1}^{M} \phi[k] - \Sigma_{k=0}^{M-1} \phi[k] + \delta[0]$$
$$= \phi[M] - \phi[0]+ \delta[0]$$
Which is better since it matches with $\int_{0}^t \delta(u)du = \int_{0}^t \frac{d\phi}{du}(u)du = \phi(t) - \phi(0)$.
Now, I think it would be better to have $\delta[0] = 0$ even if the phase is a RampWaveform (this assumes that the function was at this value before). The case of the RampWaveform is a bit tedious since there is in fact no derivative at t=0: there is one when you come from "the right" (the slope) and one when you come from "the left" (0 if you think the rampwaveform starts at 0). If you want to imagine a continuous function, then I believe the derivative should be 0 since it will be an inflection point.
Otherwise, it is possible to have $\delta[0]$ non equal to zero but I think we should modify the constant phase of the pulse accordingly since for $M=0$, $\phi[0] = \delta[0]$

Reply:
Thank you for the detailed derivation! I think your last suggestion is the best one: instead of taking $\phi_c = \phi[0]$ we can take $\phi_c = \phi[0] - \delta[0]$ and it will all add up.